File: examples/grasping.cpp
#include <frankx/frankx.hpp>


using namespace frankx;

inline Affine getBase(double x = 0.0, double y = 0.0, double z = 0.0, double a = 0.0, double b = 0.0, double c = 0.0) {
    return Affine(0.48 + x, -0.204 + y, 0.267 + z, a, b, c);
}


int main(int argc, char *argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <robot-hostname>" << std::endl;
        return -1;
    }

    Robot robot(argv[1]);
    robot.automaticErrorRecovery();
    robot.setDefaultBehavior();

    auto joint_motion = JointMotion({-1.8119446041, 1.1791089121, 1.7571002245, -2.141621800, -1.1433693913, 1.6330460616, -0.4321716643});
    auto joint_data = MotionData().withDynamicRel(0.2);
    robot.move(joint_motion, joint_data);

    auto data = MotionData().withDynamicRel(0.5).withReaction(Reaction(
        Measure::ForceZ() < -7.0,
        std::make_shared<LinearRelativeMotion>(Affine(0.0, 0.0, 0.001))
    ));

    WaypointMotion waypoint_motion({
        Waypoint {getBase(0.05, 0.05, 0.0, M_PI_2)},
        Waypoint {getBase(0.05, 0.05, -0.185, M_PI_2)}
    });
    robot.move(waypoint_motion, data);

    WaypointMotion waypoint_motion2({
        Waypoint {getBase(0.05, 0.05, 0.0, M_PI_2)},
        Waypoint {getBase(0.0, 0.0, 0.0), 1.75}
    });
    robot.move(waypoint_motion2);

    return 0;
}

File: examples/grasping.py
from dataclasses import dataclass

from frankx import *


@dataclass
class Grasp:
    """Planar grasp with stroke d"""
    x: float  # [m]
    y: float  # [m]
    z: float  # [m]
    a: float  # [m]
    d: float  # [m]


class Grasping:
    def __init__(self, host):
        self.robot = Robot(host)
        self.gripper = self.robot.get_gripper()

        self.robot.set_default_behavior()
        self.robot.set_dynamic_rel(0.2)
        self.robot.recover_from_errors()

        joint_motion = JointMotion([-1.811944, 1.179108, 1.757100, -2.14162, -1.143369, 1.633046, -0.432171])
        self.robot.move(joint_motion)

        # self.gripper.move(self.gripper.max_width)

        self.robot.move(LinearMotion(self.get_base(0, 0, 0), 1.75))

    @staticmethod
    def get_base(x, y, z, a=0.0, b=0.0, c=0.0):
        return Affine(0.48 + x, -0.204 + y, 0.37 + z, a, b, c)

    def grasp(self, grasp: Grasp):
        data_down = MotionData(0.6).with_reaction(Reaction(Measure.ForceZ < -7.0, LinearRelativeMotion(Affine(0, 0, 0))))
        data_up = MotionData(0.8)

        # self.gripper.move_async(grasp.d)

        motion_down = WaypointMotion([
            Waypoint(self.get_base(grasp.x, grasp.y, -0.04, grasp.a), 1.6),
            Waypoint(self.get_base(grasp.x, grasp.y, grasp.z, grasp.a), 1.3),
        ])

        move_success = self.robot.move(motion_down, data_down)

        if not move_success:
            self.robot.recover_from_errors()

        self.gripper.clamp()

        motion_up = WaypointMotion([
            Waypoint(self.get_base(grasp.x, grasp.y, -0.04, grasp.a), 1.6),
            Waypoint(self.get_base(0.0, 0.0, 0.0), 1.75),
        ])
        self.robot.move(motion_up, data_up)

        self.gripper.release()


if __name__ == '__main__':
    grasping = Grasping('172.16.0.2')
    grasping.grasp(Grasp(x=0.05, y=0.03, z=-0.185, a=0.6, d=0.06))  # [m]
